### ISSUE TITLE
JCE not working

### ISSUE THREAD

**Issue by userMar2017** [biemond/biemond-jdk7#21]
When I set the  entry "cryptography_extension_file => 'UnlimitedJCEPolicyJDK7.zip',"   it just does not work. Any idea on how to fix this ?

As shown below, it looks like its trying to extract to the exact same folder

Info: /Stage[main]/Safeway::Ohs::Install_jdk7/Jdk7::Install7[jdk1.7.0_91]/Jdk7::Config::Javaexec[jdkexec jdk1.7.0_91 7u91]/Exec[extract jce jdk1.7.0_91]: Scheduling refresh of Exec[Move jce jdk1.7.0_91 jar files to /opt/jdk1.7.0_91/jre/lib/security]
Notice: /Stage[main]/Safeway::Ohs::Install_jdk7/Jdk7::Install7[jdk1.7.0_91]/Jdk7::Config::Javaexec[jdkexec jdk1.7.0_91 7u91]/Exec[Move jce jdk1.7.0_91 jar files to /opt/jdk1.7.0_91/jre/lib/security]/returns: find: Failed to change directory: Permission denied
Notice: /Stage[main]/Safeway::Ohs::Install_jdk7/Jdk7::Install7[jdk1.7.0_91]/Jdk7::Config::Javaexec[jdkexec jdk1.7.0_91 7u91]/Exec[Move jce jdk1.7.0_91 jar files to /opt/jdk1.7.0_91/jre/lib/security]/returns: find: Failed to change directory: Permission denied
Notice: /Stage[main]/Safeway::Ohs::Install_jdk7/Jdk7::Install7[jdk1.7.0_91]/Jdk7::Config::Javaexec[jdkexec jdk1.7.0_91 7u91]/Exec[Move jce jdk1.7.0_91 jar files to /opt/jdk1.7.0_91/jre/lib/security]/returns: find: failed to restore initial working directory: Permission denied
Error: /Stage[main]/Safeway::Ohs::Install_jdk7/Jdk7::Install7[jdk1.7.0_91]/Jdk7::Config::Javaexec[jdkexec jdk1.7.0_91 7u91]/Exec[Move jce jdk1.7.0_91 jar files to /opt/jdk1.7.0_91/jre/lib/security]: Failed to call refresh: find /opt/jdk1.7.0_91/jre/lib/security -mindepth 2 -name '*.jar' -exec mv '{}' /opt/jdk1.7.0_91/jre/lib/security ';' returned 1 instead of one of [0]
Error: /Stage[main]/Safeway::Ohs::Install_jdk7/Jdk7::Install7[jdk1.7.0_91]/Jdk7::Config::Javaexec[jdkexec jdk1.7.0_91 7u91]/Exec[Move jce jdk1.7.0_91 jar files to /opt/jdk1.7.0_91/jre/lib/security]: find /opt/jdk1.7.0_91/jre/lib/security -mindepth 2 -name '*.jar' -exec mv '{}' /opt/jdk1.7.0_91/jre/lib/security ';' returned 1 instead of one of [0]

**Comment by biemond** [comment]
It works but it looks like you got some permissions issues. are you running puppet under root.
I mostly install it under /usr/java , can you check the permissions

see https://github.com/biemond/biemond-jdk7/blob/master/manifests/config/javaexec.pp#L45

**Comment by userMar2017** [comment]
Yes, I am running puppet as root. I install it in a different folder say "/opt"

**Comment by userMar2017** [comment]
Looks like a sequencing issue. If I let puppet run it on its own, it works. If I run it manually (using puppet agent -t), it fails.

**Comment by userMar2017** [comment]
After a lot of troubleshooting I was able to fix this by adding "cwd" to the "Move jce" block in "javaexec.pp"

Apparently, this poses a problem only if you do not use the root user. I cannot explain why.

Ed, is this something you can verify and once confirmed, probably add to the source ?

    exec { "Move jce ${full_version} jar files to ${security_dir}":
      cwd         => $security_dir,
      command     => "find ${jarfiles} -exec ${mv_cmd}",
      group       => $group,
      path        => $path,
      logoutput   => true,
      refreshonly => true,
      user        => $user,

**Comment by biemond** [comment]
Yeah I heard it before on solaris, else the operation is done from / and that requires more privs.

Thanks

**Comment by userMar2017** [comment]
Thanks for confirming Ed.

Will this be updated in the module ? I'd hate to make the change locally (on my system) only to have it overwritten the next time I update it from the source.

**Comment by biemond** [comment]
will do

**Comment by biemond** [comment]
pushed  a new version 1.0.0 and made it for puppet 4

**Comment by userMar2017** [comment]
Did some hiera stuff get added ? Its failing with following error

Error: Could not retrieve catalog from remote server: Error 500 on SERVER: Server Error: Evaluation Error: Error while evaluating a Resource Statement, Evaluation Error: Error while evaluating a Function Call, Lookup of key 'jdk7::exec_path' failed: DataBinding 'hiera': undefined method `empty?' for nil:NilClass at /etc/puppetlabs/code/environments/development/modules/jdk7/manifests/install7.pp:42:28 at /etc/puppetlabs/code/environments/development/modules/ohs/install_jdk8.pp:6 on node XXXX

**Comment by biemond** [comment]
let me check but you are using puppet 4 or not.

**Comment by userMar2017** [comment]
Puppet 4.8.2

**Comment by biemond** [comment]
ok same as me, maybe some caching or sync problem

can you check if you have this
https://github.com/biemond/biemond-jdk7/blob/puppet4/hiera.yaml
https://github.com/biemond/biemond-jdk7/blob/puppet4/data/common.yaml

and this 
https://github.com/biemond/biemond-jdk7/blob/puppet4/metadata.json#L3

**Comment by userMar2017** [comment]
That looks identical except that the version I have is 1.0.0 ; did it just get updated ? Should I download it again ?

**Comment by biemond** [comment]
no, you are fine. are you using an puppet master or got some old version somewhere

**Comment by userMar2017** [comment]
I tried this on my puppet vm and it worked ; looks like some problem with our company's puppet infrastructure. I will follow up with the Puppet admin. Very odd though as it was working fine until now. Sorry for the bother Ed.

**Comment by biemond** [comment]
no worries, mostly you have do to do some dummy master puppet run and put the module on the right locations and no duplicates.
